### PR TITLE
federated data-planes

### LINKED ISSUE #1493: federated data planes: punch list
- [x] Gazette: support `label-name:prefix` syntax for label selectors
    - Replaces existing `prefix` syntax for journal names.
    - Required for authorized Apply RPCs of shard specs.
    - Future: can power partition-based authorization over `estuary.dev/field/some-field` values.
- [x] Gazette: support streaming or long-polled List API for journal watches
    - Replaces direct Etcd watch of journals.
    - Required for cross-DP partition discover within shuffled reads and writes.
- [x] Gazette: authorization modeled as label selectors
    - JWT verification of journal read, append, list, replicate, and apply RPCs.
    - JWT verification of shard list, apply, and stat RPCs.
    - Auth checks are enabled & required if and only if started with a configured secret
    - Replaces similar authorization checks and proxying in data-plane-gateway
    - Required for securing public Gazette endpoints with unproxied peer-to-peer access.
- [x] Control plane: introduce authorization API for brokered data plane access
    - Requesting data-plane X signs claims for access from catalog name Foo/One (in X) to Bar/Two (in Y) using X's secret
        - (Note the requester doesn't need to know what data-plane the resource lives in)
    - API verifies claims and checks current RBAC
    - API responds with Y's address and scoped token for Bar/Two signed with Y's secret.
- [x] Reactors: sign and request access for read & written collections from authorization API
    - Replaces current direct access through internal broker API.
- [x] Reactors: use Gazette List API for journal watches in shuffled reads & appends
    - Replaces current direct access to Etcd journals watch
- [x] Control plane: model `data_planes` and track data-planes on `live_specs`
- [x] Control plane: updated logs / stats, L1, and L2 reporting infrastructure for federated planes
- [x] Control plane: basic automation for management of public / private data-planes
- [x] E2E Integrated testing of managed plane w/ a local stack
- [x] Refine release plane for incremental roll-out to existing control-plane & data-plane

### PR DESCRIPTION
**Pull request by jgraettinger**
**Description:**

This PR implements support for federated, public and private data-planes into the Estuary platform.

It builds on [this related Gazette PR](https://github.com/gazette/core/pull/390) which introduces fine-grained Authorizations to the Gazette RPC model (along with TLS). That work is further leveraged here to enable a new task authorization and discovery flow, whereby task shards consult the control plane to jointly a) obtain authorizations to read or written collections, and b) discover the appropriate data-plane endpoint to dispatch read / write RPCs.

We have always enforced RBAC of tasks and collections in the control plane. This new authorization flow significantly tightens how quickly a **revocation** of access takes effect and causes the runtime task to stop processing, which was not previously the case. Coupling authorization with endpoint discovery is a natural fit and allows for one data-plane to fairly seamlessly interact with authorized collections of another.

Many, _many_ related changes were all required to make this work end-to-end, which are describe in individual commits.

Resolves #1493 

**Workflow steps:**

* Create a new data-plane using a new control-plane API `/admin/create-data-plane`.
* When the data-plane is ready, add it to unified reporting via API `/admin/update-l2-reporting`.
* Assign new tasks to data-planes through a new `data_plane_name` column of `publications`.
* Tasks and collections are activated into their designated data-planes, and use new `/authorize/task` API to automatically obtain authorization tokens and discover owning data-planes.

The control-plane remembers data-plane assignments of each task or collection, and once set they cannot change without re-creating the task or collection.

**Next steps**:

This PR is completes the implementation of the core features for federated data-planes, but work remains to determine a release plan for applying it to our production environment in a staged fashion.

We also intend to move Validate and Discover RPCs into the reactor, called by the agent, but this is left as a next step until we can release these initial changes.

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/1557)

### CODE REVIEW

File: ops-catalog/flow_generated/typescript/ops/rollups/L1/BASE_NAME/catalog-stats.ts
Comment (psFried) [inline ops-catalog/flow_generated/typescript/ops/rollups/L1/BASE_NAME/catalog-stats.ts:71]: Just curious why this is nullable now?

File: ops-catalog/stats.schema.yaml
Comment (psFried) [inline ops-catalog/stats.schema.yaml:131]: `bytesTotal` seems like it's still not nullable, so not sure why it was in the TS code.

Stopping here for now. Looks great so far

**Review by psFried — COMMENTED** [review]
Not quite through it all, but looking good so far

File: ops-catalog/stats.schema.yaml
Comment (jgraettinger) [reply ops-catalog/stats.schema.yaml:131]: it's not nullable, but it's also not marked as required. Dunno when we did that, but this apparently hasn't been re-generated since.

I didn't edit this yaml -- I merely ran it through a JSON => yaml converter to make everything consistently YAML and tweaked file names, without making any content changes. Which I now kinda wish I hadn't done, but ... :shrug:

File: ops-catalog/README.md
Comment (jgraettinger) [reply ops-catalog/README.md:9]: Yea, the current migration in `supabase/migrations/` is really only suited for bootstrapping a local stack.

The work before me is to come up with an **actual** production migration which will look a bit different.
That includes given current tenants an `ops/dp/public/` grant.

File: crates/agent/src/api/authorize.rs
Comment (psFried) [inline crates/agent/src/api/authorize.rs:182]: nit: seems like an empty string here might be bad somehow. Should it return an error?

**Review by psFried — APPROVED** [review]
This is a bigun! I've finished reading the diff, and also did a bit of local testing. I think this will be good to merge once local stacks are set to allow missing authZ headers, and reactors updated to not fetch authZ tokens when auth is disabled.

File: supabase/seed.sql
Comment (jgraettinger) [reply supabase/seed.sql:73]: Pinged you on Slack. tl;dr is that's what's currently asserted and I perpetuated that: 
https://github.com/estuary/flow/blob/master/supabase/seed.sql#L51-L52

Would be nice to back this out of course.

File: crates/agent/src/integration_tests/mod.rs
Comment (psFried) [inline crates/agent/src/integration_tests/mod.rs:15]: Didn't catch this earlier, but I'm changing that test a bunch anyway, and can remove this while I'm at it.

**Comment by jgraettinger** [comment]
Okay, I've pushed up a last commit I intend to squash into other commits before merging. See commit message for what's changed.

I've run the release plan worked up in [this comment](https://github.com/estuary/flow/issues/1493#issuecomment-2312947724) against a local stack a few times, caught real issues along the way, and am as confident as I think I can be.

Assuming it looks good, I'd love to land and start executing on the release plan tomorrow

**Review by psFried — APPROVED** [review]
LGTM :rocket:

**Pull request merged**
